src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -416,62 +345,163 @@ TryToIncrementSharedConnectionCounter(const char *hostname, int port)
+IncrementSharedConnectionCounterInternal(uint32 externalFlags,
+                                         bool checkLimits,
+                                         const char *hostname,
+                                         int port,
+                                         Oid database)
+{
+    LockConnectionSharedMemory(LW_EXCLUSIVE);
+
+    /*
+     * As the hash map is allocated in shared memory, it doesn't rely on palloc for
+     * memory allocation, so we could get NULL via HASH_ENTER_NULL when there is no
+     * space in the shared memory. That's why we prefer continuing the execution
+     * instead of throwing an error.
+     */
+    SharedWorkerNodeConnStatsHashKey workerNodeKey = PrepareWorkerNodeHashKey(hostname, port);
+    bool workerNodeEntryFound = false;
+    SharedWorkerNodeConnStatsHashEntry *workerNodeConnectionEntry =
+            hash_search(SharedWorkerNodeConnStatsHash,
+                        &workerNodeKey,
+                        HASH_ENTER_NULL,
+                        &workerNodeEntryFound);
+
+    /*
+     * It is possible to throw an error at this point, but that doesn't help us in anyway.
+     * Instead, we try our best, let the connection establishment continue by-passing the
+     * connection throttling.
+     */
+    if (!workerNodeConnectionEntry)
+    {
+        UnLockConnectionSharedMemory();
+        return true;
+    }
+
+    if (!workerNodeEntryFound)
+    {
+        /* we successfully allocated the entry for the first time, so initialize it */
+        workerNodeConnectionEntry->regularConnectionsCount = 0;
+        workerNodeConnectionEntry->maintenanceConnectionsCount = 0;
+    }
+
+    /* Initialize SharedWorkerNodeDatabaseConnStatsHash the same way  */
+    SharedWorkerNodeDatabaseConnStatsHashKey workerNodeDatabaseKey =
+            PrepareWorkerNodeDatabaseHashKey(hostname, port, database);
+    bool workerNodeDatabaseEntryFound = false;
+    SharedWorkerNodeDatabaseConnStatsHashEntry *workerNodeDatabaseEntry =
+            hash_search(SharedWorkerNodeDatabaseConnStatsHash,
+                        &workerNodeDatabaseKey,
+                        HASH_ENTER_NULL,
+                        &workerNodeDatabaseEntryFound);
+
+    if (!workerNodeDatabaseEntry)
+    {
+        UnLockConnectionSharedMemory();
+        return true;
+    }
+
+    if (!workerNodeDatabaseEntryFound)
+    {
+        workerNodeDatabaseEntry->count = 0;
+    }
+
+    /* Increment counter if a slot available */
+    bool connectionSlotAvailable = true;
+
+    bool maintenanceConnection = externalFlags & MAINTENANCE_CONNECTION;
+    if (checkLimits)
+    {
+        WorkerNode *workerNode = FindWorkerNode(hostname, port);
+        bool connectionToLocalNode = workerNode && (workerNode->groupId == GetLocalGroupId());
+
+        int currentConnectionsLimit;
+        int currentConnectionsCount;
+        if (maintenanceConnection)
+        {
+            currentConnectionsLimit = GetMaxMaintenanceSharedPoolSize();
```

Comment:
@marcoslot, not sure about this moment: current implementation has common `MaxMaintenanceSharedPoolSize` both for remote and local connections, because it is simpler. Is it necessary to introduce separate limit for local maintenance connections?
Another effect of this is that when `GetLocalSharedPoolSize() == DISABLE_REMOTE_CONNECTIONS_FOR_LOCAL_QUERIES`, remote connections are disabled both for maintenance and regular connections